lnst/Recipes/ENRT/ConfigMixins/LongLivedConnectionsMixin.py

```diff
@@ -0,0 +1,230 @@
+    connections, while the last one will get the remaining connections.
+
+    Don't forget to set appropriate NO_FILES ulimit (if needed).
+    See LongLivedServer/LongLivedClient for more details.
+    """
+
+    long_lived_conns = IntParam(mandatory=True)
+    long_lived_conns_port = IntParam(default=20000)
+    long_lived_conns_per_ip = IntParam(default=20000)
+
+    def _generate_ip_endpoints(self, config):
+        return [
+            ip_endpoint_pairs(
+                config,
+                (self.matched.host1.eth0, self.matched.host2.eth0),
+                combination_func=zip,
+            )
+        ]
+
+    def _filter_ip_endpoints(self, config, slicer):
+        ips = []
+
+        for parallel_endpoints in self._generate_ip_endpoints(config):
+            ipv4 = [
+                endpoint_pair
+                for endpoint_pair in parallel_endpoints
+                if isinstance(endpoint_pair.first.address, Ip4Address)
+            ]
+            ipv6 = [
+                endpoint_pair
+                for endpoint_pair in parallel_endpoints
+                if isinstance(endpoint_pair.first.address, Ip6Address)
+            ]
+
+            ips.append(ipv4[slicer])
+            ips.append(ipv6[slicer])
+
+        return ips
+
+    def generate_perf_endpoints(self, config):
+        return self._filter_ip_endpoints(
+            config, slice(0, 1)  # only the first IP is used for perf
+        )
+
+    def generate_long_lived_conns_endpoints(self, config):
+        return self._filter_ip_endpoints(
+            config,
+            slice(1, None),  # all IPs except the first one which is used for perf
+        )
+
+    @property
+    def servers_count(self):
+        return ceil(self.params.long_lived_conns / self.params.long_lived_conns_per_ip)
+
+    def calculate_client_connections(self, client_id):
+        client_id += 1  # 0-based index
+        if client_id < self.servers_count:
+            return self.params.long_lived_conns_per_ip
+
+        # remaining connections are handled by last client
+        return (
+            self.params.long_lived_conns
+            - (self.servers_count - 1) * self.params.long_lived_conns_per_ip
+        )
+
+    def _prepare_server(self, receiver, conns_count):
+        server = LongLivedServer(
+            server_ip=receiver.address,
+            server_port=self.params.long_lived_conns_port,
+            connections_count=conns_count,
+        )
+
+        job = receiver.device.netns.prepare_job(server)
+
+        return job
+
+    def _prepare_client(self, generator, receiver, conns_count):
+        client = LongLivedClient(
+            server_ip=receiver.address,
+            server_port=self.params.long_lived_conns_port,
+            client_ip=generator.address,
+            connections_count=conns_count,
+        )
+
+        job = generator.device.netns.prepare_job(client)
+
+        return job
+
+    def wait_agent_for_establish_conns(self, agent, timeout=60):
```

Comment:
discussed live in a tech meeting, IMO you should be able to implement a Test module for this instead:
```
agent.run(ConnectionsEstablished(num_of_connections), timeout=timeout)
```
where
```
class ConnectionsEstablished(BaseTestModule):
     num_of_connections = IntParam
     def run(self):
        while True:
            stdout = exec(ss | ESTAB | wc -l)
            if int(stdout) >= num_of_connections:
               return True
```

Reply:
Yep, I did something similar to this but more generic, see `Agent: waiting for condition on agent side` and `Agent/Conditions: Waiting for connections establishment` commits

Reply:
ack, made new comments on those, resolving this older conversation...